### ISSUE TITLE
Occasionally Source salt://xxx.tar.gz not found 

### ISSUE THREAD

**Issue by mosuowhq** [saltstack/salt#29656]
just like #10878
Minion number is over 60.  When 2 or more minions fetch the same file, "Source file salt://xxxx.tgz not found" occurs in one minion. 
minion log with error is as below:
```
2015-12-10 22:18:52,522 [salt.state                                 ][INFO    ][2445] Running state [/path/XXX.tar.gz] at time 22:18:52.522171
2015-12-10 22:18:52,522 [salt.state                                 ][INFO    ][2445] Executing state file.managed for /path/XXX.tar.gz
2015-12-10 22:18:53,650 [salt.state                                 ][ERROR   ][2445] Source file salt://XXX.tar.gz not found
2015-12-10 22:18:53,651 [salt.state                                 ][INFO    ][2445] Completed state [/path/XXX.tar.gz] at time 22:18:53.650625
```
But the file XXX.tar.gz actually exists in the salt:// and other minions download the files successfully.
```yaml
versions-report:
                  Salt: 2015.5.3
                Python: 2.7.6 (default, Mar 22 2014, 22:59:56)
                Jinja2: 2.7.2
              M2Crypto: 0.21.1
        msgpack-python: 0.3.0
          msgpack-pure: Not Installed
              pycrypto: 2.6.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 14.0.1
                  RAET: Not Installed
                   ZMQ: 4.0.4
                  Mako: Not Installed
               Tornado: Not Installed
```

**Comment by jfindlay** [comment]
@mosuowhq, thanks for the report.  If you find any more information that may help identify the circumstances or causes of this issue, please add it here.

**Comment by mosuowhq** [comment]
@jfindlay
The error occurs again and the master log reports as below:

2015-12-17 21:57:10,524 [salt.master      ][ERROR   ][18403] Error in function _file_hash:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/master.py", line 1457, in run_func
    ret = getattr(self, func)(load)
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/__init__.py", line 528, in file_hash
    return self.servers[fstr](load, fnd)
  File "/usr/lib/python2.7/dist-packages/salt/fileserver/roots.py", line 233, in file_hash
    os.makedirs(cache_dir)
  File "/usr/lib/python2.7/os.py", line 157, in makedirs
    mkdir(name, mode)
OSError: [Errno 17] File exists: '/var/cache/salt/master/roots/hash/base/package//common'
2015-12-17 21:57:10,537 [salt.utils.event ][DEBUG   ][18424] Sending event - data = {'jid': '20151217215653847724', 'cmd': '_minion_event', '_stamp': '2015-12-17T13:57:10.536419', 'tag': 'salt/job/20151217215653847724/prog/cloudify1/44', 'data': {'ret': {'comment': 'Source file salt://package//common/haproxy.tgz not found', 'name': '/var/cache/salt/minion/files/base/_var_paas_jobs_haproxy_.tar', 'start_time': '21:57:09.311488', 'result': False, 'duration': 973.907, '__run_num__': 44, 'changes': {}}, 'len': 208}, 'id': 'cloudify1'}

It seems that the cache_dir has been exist when executing os.makedirs(cache_dir).
I modified /usr/lib/python2.7/dist-packages/salt/fileserver/roots.py, line 233. 
Is there any potential risk?

    # make cache directory if it doesn't exist
    if not os.path.exists(cache_dir):
        try:
            os.makedirs(cache_dir)
        except OSError:
            log.warning("Fileserver failed to make cache directory. Retrying.")
            return file_hash(load, fnd)

**Comment by jfindlay** [comment]
@mosuowhq, thanks for the extra info.  This should be fixed by #28545, which is available in 2015.8.3.

**Comment by mosuowhq** [comment]
@jfindlay thanks for you response, i will close this issue.
